pkg/networkservice/common/heal/server.go

```diff
@@ -356,43 +350,38 @@ func (f *healServer) restoreConnection(ctx, baseCtx context.Context, request *ne
 	if deadline.After(expireTime) {
 		deadline = expireTime
 	}
-	requestCtx, requestCancel := context.WithDeadline(f.ctx, deadline)
+	requestCtx, requestCancel := context.WithDeadline(ctx, deadline)
 	defer requestCancel()
 
-	for f.ctx.Err() == nil {
-		_, err = (*f.onHeal).Request(requestCtx, request.Clone(), opts...)
-		if healMapCtx != ctx {
+	for requestCtx.Err() == nil {
+		if _, err = (*f.onHeal).Request(requestCtx, request.Clone(), opts...); err == nil {
 			return
 		}
-		if time.Now().After(deadline) || err == nil {
-			break
-		}
 	}
 
 	if err != nil {
-		f.processHeal(baseCtx, request.Clone(), opts...)
+		f.processHeal(ctx, baseCtx, request.Clone(), opts...)
 	}
 }
 
-func (f *healServer) processHeal(baseCtx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) {
-	logEntry := log.FromContext(f.ctx).WithField("healServer", "processHeal")
+func (f *healServer) processHeal(ctx, baseCtx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) {
```

Comment:
Let's just pass one heal context which will include candidates :) 
```suggestion
func (f *healServer) processHeal(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) {
```

Reply:
Yep, fixed.